src/lib.rs

```diff
@@ -0,0 +1,162 @@
+
+        let serenity_client =
+            SerenityClient::builder(config.discord_access_token.expose_secret(), intents)
+                .event_handler(Handler)
+                .await?;
+        let serenity_data = serenity_client.data.clone();
+        {
+            let mut data = serenity_data.write().await;
+            data.insert::<CalendarClient>(calendar_client.clone());
+            data.insert::<Pool>(pool);
+        }
+
+        let discalen_client = Self {
+            discord_client: DiscordClient::new(serenity_client).await,
+        };
+        let discord_client = discalen_client.discord_client;
+        let sender_http = discord_client.serenity_client.http.clone();
+        let discord_data = discord_client.serenity_client.data.clone();
+
+        let discord_task = tokio::spawn(async move {
+            let mut serenity_client = discord_client.serenity_client;
+            if let Err(why) = serenity_client.start_autosharded().await {
+                println!("Client error: {why:?}");
+            }
+        });
+
+        let calendar_task: JoinHandle<Result<(), Error>> = tokio::spawn(async move {
+            let calendar_client = calendar_client;
+            loop {
+                tokio::time::sleep(config.notification_period).await;
+
+                let calendars = calendar_client.list_calendars().await?;
+
+                let mut calendars_handles = vec![];
+                for calendar in &calendars {
+                    let handle = calendar_client
+                        .list_events(calendar.id.as_ref().unwrap_or_else(|| unreachable!()));
+                    calendars_handles.push((calendar, handle));
+                }
+
+                let mut stream = futures::stream::iter(calendars_handles);
+                while let Some(handle) = stream.next().await {
+                    let (calendar, events) = handle;
+                    let events = events.await?;
+
+                    let mut sending_tasks = vec![];
+                    for event in events {
+                        let date = match event.start {
+                            Some(EventDateTime {
+                                date: Some(date), ..
+                            }) => date,
+                            _ => {
+                                warn!(?event, "The event doesn't have the start date, skipping...");
+                                continue;
+                            }
+                        };
+                        if date == Utc::now().date_naive() {
+                            sending_tasks.push(send_event_notification(
+                                discord_data.clone(),
+                                &sender_http,
+                                calendar,
+                                event,
+                            ));
+                        }
+                    }
+                    futures::future::join_all(sending_tasks).await;
```

Comment:
https://docs.rs/tokio/latest/tokio/task/struct.JoinSet.html

Reply:
Thanks for the mention of `JoinSet`, but unfortunately, they introduce a lot of complexity due to lifetimes. Couldn't fix it in 2 hours, so I think I'll stick to `join_all` from `futures` for now

Reply:
Valid.